helix-gateway/src/main/java/org/apache/helix/gateway/util/StateTransitionMessageTranslateUtil.java

```diff
@@ -1,17 +1,70 @@
 package org.apache.helix.gateway.util;
 
-import org.apache.helix.gateway.service.GatewayServiceManager;
-import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass.TransitionMessage;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.helix.gateway.constant.GatewayServiceEventType;
+import org.apache.helix.gateway.service.GatewayServiceEvent;
+import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass;
+import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass.ShardState;
 import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass.ShardStateMessage;
+import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass.ShardTransitionStatus;
+import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass.TransitionMessage;
 
 
 public final class StateTransitionMessageTranslateUtil {
 
-  public static TransitionMessage translateSTMsgToProto() {
+  public static TransitionMessage translateSTMsgToTransitionMessage() {
     return null;
   }
 
-  public static GatewayServiceManager.GateWayServiceEvent translateProtoToSTMsg(ShardStateMessage message) {
-    return null;
+  /**
+   * Translate from user sent ShardStateMessage message to Helix Gateway Service event.
+   */
+  public static GatewayServiceEvent translateShardStateMessageToEvent(ShardStateMessage request) {
+
+    GatewayServiceEvent.GateWayServiceEventBuilder builder;
+    if (request.hasShardState()) { // init connection to gateway service
+      ShardState shardState = request.getShardState();
+      Map<String, String> shardStateMap = new HashMap<>();
+      for (HelixGatewayServiceOuterClass.SingleResourceState resourceState : shardState.getResourceStateList()) {
+        for (HelixGatewayServiceOuterClass.SingleShardState state : resourceState.getShardStatesList()) {
+          shardStateMap.put(resourceState.getResource() + "_" + state.getShardName(), state.getCurrentState());
+        }
+      }
+      builder = new GatewayServiceEvent.GateWayServiceEventBuilder(GatewayServiceEventType.CONNECT).setClusterName(
+          shardState.getClusterName()).setParticipantName(shardState.getInstanceName());
+    } else {
+      ShardTransitionStatus shardTransitionStatus = request.getShardTransitionStatus();
+      // this is status update for established connection
+      List<HelixGatewayServiceOuterClass.SingleShardTransitionStatus> status =
+          shardTransitionStatus.getShardTransitionStatusList();
+      List<GatewayServiceEvent.StateTransitionResult> stResult = new ArrayList<>();
+      for (HelixGatewayServiceOuterClass.SingleShardTransitionStatus shardTransition : status) {
+        String stateTransitionId = shardTransition.getTransitionID();
+        String stateTransitionStatus = shardTransition.getCurrentState();
+        String shardState = shardTransition.getCurrentState();
```

Comment:
As I said, no creation for these variables when it only used once...

Reply:
Updated.